network/src/commonMain/kotlin/com/wire/kalium/network/exceptions/KaliumException.kt

```diff
@@ -1,27 +1,16 @@
 package com.wire.kalium.network.exceptions
 
-import kotlinx.serialization.Serializable
+import com.wire.kalium.network.api.ErrorResponse
+import com.wire.kalium.network.api.message.SendMessageResponse
 
-// TODO: return custom error instead of throwing exception
-
-@Serializable
-sealed class KaliumException : Exception()
-
-@Serializable
-sealed class NetworkException(
-    open var _message: String? = null,
-    open var _code: Int = 0,
-    open var _label: String? = null
-) : KaliumException()
-
-@Serializable
-class AuthException : NetworkException {
-    constructor(message: String?, code: Int) : super(_message = message, _code = code)
-    constructor(code: Int) : super(_code = code)
+sealed class KaliumException(val errorCode: Int) : Exception() {
+    class RedirectError(val errorResponse: ErrorResponse, throwable: Throwable?) : KaliumException(errorCode = errorResponse.code)
+    class InvalidRequestError(val errorResponse: ErrorResponse, throwable: Throwable?) : KaliumException(errorCode = errorResponse.code)
+    class ServerError(val errorResponse: ErrorResponse, throwable: Throwable?) : KaliumException(errorCode = errorResponse.code)
+    class GenericError(val errorResponse: ErrorResponse?, throwable: Throwable?) : KaliumException(errorCode = errorResponse?.code ?: 400)
```

Comment:
These `Throwable`s are not being used.

For non-generic errors, like lack of connection, 
I don't think they are actually needed for non-generic errors.

But for Generic ones, I guess they are always caused by an underlying exception, so we should always add the throwable as cause and don't accept a `null` cause. Generic errors can be bad on their own, at least we need to make sure that we can know what caused them.

So, my suggestion:

```suggestion
sealed class KaliumException(val errorCode: Int) : Exception() {
    val errorCode: Int

    constructor(errorCode: Int) : super() {
        this.errorCode = errorCode
    }
    constructor(errorCode: Int, cause: Throwable) : super(cause) {
        this.errorCode = errorCode
    }

    class RedirectError(val errorResponse: ErrorResponse) : KaliumException(errorCode = errorResponse.code)
    class InvalidRequestError(val errorResponse: ErrorResponse) : KaliumException(errorCode = errorResponse.code)
    class ServerError(val errorResponse: ErrorResponse) : KaliumException(errorCode = errorResponse.code)
    class GenericError(val errorResponse: ErrorResponse?, cause: Throwable) : KaliumException(errorCode = errorResponse?.code ?: 400, cause = cause)
```

Reply:
Yep, sounds about right. I didn't want to get too extensive on this, but good point on taking our time now. Let me know if it looks good enough now